File: swift/CarDriveMAIN.playground/Sources/GameObjects/Tree.swift
import Foundation

public class Tree: EnvironmentObject {
    var radius: Double

    init(x: Double, y: Double, radius: Double = 1.0) {
        self.radius = radius
        super.init(x: x, y: y, name: "tree")
    }

    // SAT collision between rotated rectangle (player) and circle (tree)
    override func checkCollisionWithPlayerDetailed(_ player: VehicleCollisionObject) -> Bool {
        let angle = player.steeringAngle * .pi / 180
        let c = cos(angle), s = sin(angle)
        let hw = player.width / 2, hl = player.length / 2

        let localCorners: [(Double, Double)] = [(-hw, -hl), (hw, -hl), (hw, hl), (-hw, hl)]
        let wc = localCorners.map { (lx, ly) -> (Double, Double) in
            (player.x + lx * c - ly * s,
             player.y + lx * s + ly * c)
        }

        // Step 1: check edge normals as separating axes
        for i in 0..<4 {
            let j = (i + 1) % 4
            let ex = wc[j].0 - wc[i].0
            let ey = wc[j].1 - wc[i].1
            let len = sqrt(ex * ex + ey * ey)
            guard len > 0 else { continue }
            let nx = -ey / len, ny = ex / len

            var minR = Double.infinity, maxR = -Double.infinity
            for corner in wc {
                let p = corner.0 * nx + corner.1 * ny
                minR = min(minR, p); maxR = max(maxR, p)
            }
            let cp = x * nx + y * ny
            if maxR < cp - radius || minR > cp + radius { return false }
        }

        // Step 2: axis from circle centre to closest point on rectangle
        var closestX = x, closestY = y
        for i in 0..<4 {
            let j = (i + 1) % 4
            let ex = wc[j].0 - wc[i].0
            let ey = wc[j].1 - wc[i].1
            let tx = x - wc[i].0, ty = y - wc[i].1
            let el2 = ex * ex + ey * ey
            guard el2 > 0 else { continue }
            let t = max(0, min(1, (tx * ex + ty * ey) / el2))
            let cx = wc[i].0 + t * ex, cy = wc[i].1 + t * ey
            let d1 = (x - cx) * (x - cx) + (y - cy) * (y - cy)
            let d2 = (x - closestX) * (x - closestX) + (y - closestY) * (y - closestY)
            if d1 < d2 { closestX = cx; closestY = cy }
        }

        let dx = x - closestX, dy = y - closestY
        return dx * dx + dy * dy < radius * radius
    }

    override func getCollisionObject() -> EnvironmentObject { self }
}

File: WebVersion/src/Tree.ts
import { Player } from './Player.js';
import { EnvironmentObject } from './EnvironmentObjects.js';
import { VehicleCollisionObject } from './Vehicle.js';

export class Tree extends EnvironmentObject {

    radius: number = 1;

    constructor(x: number, y: number, radius: number = 1) {
        super(x, y, "🌳");

        this.radius = radius;
    }


    // TBC - this could be a generic method for all environment objects (e.g. trees, vehicles, etc.)
    checkCollisionWithPlayerDetailed(player: VehicleCollisionObject): boolean {
        // SAT (Separating Axis Theorem) collision detection between rotated rectangle and circle
        // https://www.sevenson.com.au/programming/sat/

        // Step 1: Calculate the four corners of the rotated rectangle
        const angle = player.steeringAngle * Math.PI / 180; // Convert to radians
        const cos = Math.cos(angle);
        const sin = Math.sin(angle);

        const halfWidth = player.width / 2;
        const halfLength = player.length / 2;

        // Local corners (before rotation, relative to center)
        const localCorners = [
            { x: -halfWidth, y: -halfLength },
            { x: halfWidth, y: -halfLength },
            { x: halfWidth, y: halfLength },
            { x: -halfWidth, y: halfLength }
        ];

        // Rotate and translate corners to world coordinates
        const worldCorners = localCorners.map(corner => ({
            x: player.x + corner.x * cos - corner.y * sin,
            y: player.y + corner.x * sin + corner.y * cos
        }));

        // Step 2: Get edge vectors and their normals
        // const edges: Array<{ x: number; y: number }> = [];
        const normals: Array<{ x: number; y: number }> = [];

        for (let i = 0; i < 4; i++) {
            const next = (i + 1) % 4;
            const edge = {
                x: worldCorners[next].x - worldCorners[i].x,
                y: worldCorners[next].y - worldCorners[i].y
            };
            // edges.push(edge);

            // Normalize edge to get normal
            const length = Math.sqrt(edge.x * edge.x + edge.y * edge.y);
            if (length > 0) {
                // Perpendicular vector (normal) - swap x/y and negate one
                normals.push({
                    x: -edge.y / length,
                    y: edge.x / length
                });
            }
        }

        // Step 3: Check each separating axis
        for (const normal of normals) {
            // Project rectangle corners onto the axis
            let minRect = Infinity;
            let maxRect = -Infinity;

            for (const corner of worldCorners) {
                const projection = corner.x * normal.x + corner.y * normal.y; // dot product of corner and the normal
                minRect = Math.min(minRect, projection);
                maxRect = Math.max(maxRect, projection);
            }

            // Project circle onto the axis
            const circleProjection = this.x * normal.x + this.y * normal.y;
            const minCircle = circleProjection - this.radius;
            const maxCircle = circleProjection + this.radius;

            // Check for separation (gap between projections)
            if (maxRect < minCircle || minRect > maxCircle) {
                // Found a separating axis - no collision
                return false;
            }
        }

        // Step 4: Check axis from circle center to closest point on rectangle
        // Find closest point on rectangle to circle center
        let closestX = this.x;
        let closestY = this.y;

        // Clamp circle center to rectangle bounds
        for (let i = 0; i < 4; i++) {
            const next = (i + 1) % 4;
            const edge = {
                x: worldCorners[next].x - worldCorners[i].x,
                y: worldCorners[next].y - worldCorners[i].y
            };

            const toPoint = {
                x: this.x - worldCorners[i].x,
                y: this.y - worldCorners[i].y
            };

            const edgeLength = edge.x * edge.x + edge.y * edge.y;
            if (edgeLength > 0) {
                const t = Math.max(0, Math.min(1, (toPoint.x * edge.x + toPoint.y * edge.y) / edgeLength));
                const closestOnEdge = {
                    x: worldCorners[i].x + t * edge.x,
                    y: worldCorners[i].y + t * edge.y
                };

                const distToEdge = Math.sqrt(
                    (this.x - closestOnEdge.x) ** 2 + (this.y - closestOnEdge.y) ** 2
                );
                const distToClosest = Math.sqrt(
                    (this.x - closestX) ** 2 + (this.y - closestY) ** 2
                );

                if (distToEdge < distToClosest) {
                    closestX = closestOnEdge.x;
                    closestY = closestOnEdge.y;
                }
            }
        }

        // Check distance from circle to closest point
        const dx = this.x - closestX;
        const dy = this.y - closestY;
        const distance = Math.sqrt(dx * dx + dy * dy);

        // If distance is less than radius, there's a collision
        return distance < this.radius;
    }
}